README.md
# Paylike's Flutter engine

[![pub points](https://badges.bar/paylike_flutter_engine/pub%20points)](https://pub.dev/packages/paylike_flutter_engine)

This library includes the core elements required to implement a payment flow towards the Paylike API.

If you are looking for our high level component providing complete payment forms as well, [check here](https://github.com/paylike/flutter-payment-forms)

## Table of contents
* [API Reference](#api-reference)
* [PaylikeEngineWidget](#paylikeenginewidget) (Webview component)
  * [Box constraints](#box-constraints)
  * [Understanding TDS](#understanding-tds)
* [PaylikeEngine](#paylikeengine) (Underlying business logic service)
  * [Engine events](#engine-events)
* [Example application](#example-application)

## API Reference

For the library you can find the API reference [here](https://paylike.io#todo-link).

To get more familier with our server API you can find here the [official documentation](https:/github.com/paylike/api-reference).

## PaylikeEngineWidget

Webview component of the payment flow, able to render the webview required to execute the TDS challenge.

Simple usage:
```dart
    /// Renders the webview component automatically when the engine needs it, otherwise
    /// shows an empty state
    Row(mainAxisAlignment: MainAxisAlignment.center, children: [
      PaylikeEngineWidget(engine: _engine, showEmptyState: true)
    ])
```

For a realistic usage check out the example [here](./example/lib/main.dart).

#### Box constraints

Currently the webview component supports both unlimited and limited constraints.
This is achieved using SizedBoxes: 

```dart
     if (constraints.maxHeight == double.infinity &&
          constraints.maxWidth == double.infinity) {
        return Center(
            child: SingleChildScrollView(
                child: SizedBox(
                    child: webviewContent,
                    height: 400,
                    width: MediaQuery.of(context).size.width - 150)));
      }
      if (constraints.maxHeight == double.infinity) {
        return SizedBox(height: 400, child: Expanded(child: webviewContent));
      }
      if (constraints.maxWidth == double.infinity) {
        return SizedBox(
            width: MediaQuery.of(context).size.width - 150,
            child: Expanded(child: webviewContent));
      }
      return SizedBox.expand(child: webviewContent);
```

If both constraints are set to infinity, the webview will be rendered in a scrollable view.

If only one of the constraints is set to infinity, the webview will be rendered in a fixed size.

Otherwise if the constraints are set to a specific value, the webview will be rendered in a fixed size.

#### Understanding TDS

TDS is required to execute the payment flow and it is a core part of accepting payments online. Every bank is required by financial laws to provide this methodology for their customers in order to achieve higher security measures.

Apple Pay is an exception from this as their system is already integrated with the TDS challenge.

<img src="https://i.imgur.com/51Ix8AP.jpg" style="max-width: 100%;"/>

## PaylikeEngine

The core component of the payment flow.

Essentially designed to be event based to allow as much flexibility as possible on the implementer side.

Simple usage:
```dart
import 'package:paylike_flutter_engine/paylike_flutter_engine.dart';

void main() async {
  var engine = PaylikeEngine(clientId: 'your-client-id');

  engine.addListener(() {
    if (engine.current == PaylikeEngineStates.webviewChallengeRequired) {
      /// show webview and continue solving the iframe challenge
    }
    /// etc...
  });

  /// Simple card usage
  /// -----------------
  /// After the createPayment function is called the engine will update its state
  /// to render TDS webview challenge
  /// NOTE: Tokenized cards can be saved and reused later with the API
  {
    var tokenizedCard = await engine.tokenize('4100000000000000', '000');
    await _engine.createPayment(CardPayment(
        card: PaylikeCard(
            details: card, expiry: const Expiry(year: 2025, month: 3)),
        amount: Money.fromDouble('EUR', 20.5),
      ));
  }


  /// Apple Pay
  /// --------
  /// Apple Pay supports TDS by default so if everything is in order you should never see
  /// engine state updates here that requires a webview to render, therefore you can only listen to
  /// 3 events: waitingForInput, errorHappened, done
  {
    var applePayToken = ''; /// Acquire token using ApplePay API or use our higher level SDK: flutter-payment-forms
    var tokenizedApplePay = await engine.tokenizeApplePay(applePayToken);
    await _engine.createPaymentWithApple(ApplePayPayment(
        token: tokenized,
        amount: Money.fromDouble('HUF', 150.0)));
  }
}
```

For further information on how the engine can be integrated into the application, please see the source of [PaylikeEngineWidget](./lib/paylike_flutter_engine/paylike_engine_widget.dart) and the library's API Reference.

#### Engine events

The library exposes an enum called EngineStates which describes the following states:

* `waitingForInput` - Indicates that the engine is ready to be used and waiting for input
* `webviewChallengeRequired` -  Indicates that a webview challenge is required to complete the TDS flow, this is the first state when the webview has to be rendered
* `webviewChallengeStarted` - Indicates that the first step of the TDS flow is done and the challenge needs interraction from the end user to resolve
* `done` - Indicates that all challenges are done successfully and the payment is being processed
* `errorHappened` - Happens when the flow could not be completed successfully

Learn more about this in the [API Reference](https://paylike.io#todo-link)

## Example application

In the [example directory](./example/lib/main.dart) you can find a simple example of how to use the library.

On both **iOS** and **Android** you need to [register a merchant account](https://paylike.io/sign-up) with Paylike before you can use the example application. Once you have created your account you can create a new client ID for yourself and use it in the sandbox environment.

**Android**

Enter your client ID to the example's main.dart, and you are good to go with running the application.

The minimum supported SDK is 19.

**iOS**

Apple Pay is only available for iOS devices. To be able to use this service you have to make a couple of modifications and [register a merchant identification with Apple](https://developer.apple.com/library/archive/ApplePay_Guide/Configuration.html).

Once you have acquired the merchant identification from Apple, you can request a payment processing certificate to be added to your account via our [support](https://paylike.io/contact). This is vital to use our API with Apple Pay.

When you received your Payment Processing Certificate, you need to add it to the [App Store Connect](https://appstoreconnect.apple.com/). Once you are done with this your merchant will be able to execute payments with Apple Pay.

In your Xcode you also have to set your own merchant identification when you are building the applicaiton. You can find this option under the Signing & Capabilities (by opening `example/ios/Runner.xcworkspace` and adding Apple Pay capability)

Now you should be ready to run the example and execute payments using Apple Pay.

The minimum supported iOS version is 11.


CHANGELOG.md
## 0.1.2

* Upgraded `paylike_money` dependency to `1.0.3`
* Removed `paylike_currencies` dependency due to deprecation

## 0.1.1

* Upgraded `paylike_money` dependency to `1.0.2`

## 0.1.0

* Alpha release of the library with initial core functionality


example/lib/main.dart
// ignore_for_file: avoid_print

import 'package:flutter/material.dart';
import 'package:pay/pay.dart';
import 'package:paylike_flutter_engine/domain.dart';
import 'package:paylike_flutter_engine/exception.dart';
import 'package:paylike_flutter_engine/paylike_flutter_engine.dart';

void main() {
  runApp(const MyApp());
}

/// Register your own account [here](https://paylike.io)
var clientID = 'your-client-id';

const _paymentItems = [
  PaymentItem(
    label: 'Total',
    amount: '150',
    status: PaymentItemStatus.final_price,
  )
];

/// This is the same example application you would get with
/// flutter create -t app .
class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Paylike Engine Usage Demo',
      theme: ThemeData(
        primarySwatch: Colors.green,
      ),
      home: const MyHomePage(title: 'Paylike Engine Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({Key? key, required this.title}) : super(key: key);
  final String title;
  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  String transactionId = "";
  String error = "";
  final PaylikeEngine _engine = PaylikeEngine(clientId: clientID);

  /// Used for listening to engine events
  void _engineListener() {
    switch (_engine.current) {
      case EngineState.errorHappened:
        setState(() {
          error = _engine.error!.message;
        });
        break;
      case EngineState.done:
        setState(() {
          transactionId = _engine.transactionId;
        });
        break;
      default:
        break;
    }
  }

  /// Start listening to [PaylikeEngine] events in the initState
  @override
  void initState() {
    super.initState();
    _engine.addListener(_engineListener);
  }

  /// Stop listening to [PaylikeEngine] events in dispose
  ///
  /// Or stop listening when you don't need it anymore
  @override
  void dispose() {
    super.dispose();
    _engine.removeListener(_engineListener);
  }

  /// Important for the Engine usage example executing a single
  /// card transaction
  void _doPaymentFlow() async {
    _engine.restart();
    var card = await _engine.tokenize("410000000000000", "123");
    await _engine.createPayment(CardPayment(
      card: PaylikeCard(
          details: card, expiry: const Expiry(year: 2025, month: 3)),
      amount: Money.fromDouble('EUR', 20.5),
    ));
  }

  /// Important for the Engine usage example executing a single
  /// Apple Pay transaction
  ///
  /// NOTE: You need to have Apple Merchant set up. Check our documentation for
  /// further information!
  void onApplePayResult(Map<String, dynamic> paymentResult) async {
    _engine.restart();
    try {
      var token = paymentResult['token'];
      var tokenized = await _engine.tokenizeAppleToken(token);
      await _engine.createPaymentWithApple(ApplePayPayment(
          token: tokenized, amount: Money.fromDouble('HUF', 150.0)));
    } on PaylikeException catch (e) {
      print(e.cause);
      print(e.code);
      print(e.statusCode);
    }
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Center(
        child: SingleChildScrollView(
            child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            const Text(
              'Press button to start payment flow',
            ),
            Text(
              'TransactionID: $transactionId',
            ),
            Text(
              'Error: $error',
            ),

            /// Notice that the widget is always rendered, but only visible
            /// when the webview flow is being done
            Row(mainAxisAlignment: MainAxisAlignment.center, children: [
              PaylikeEngineWidget(engine: _engine, showEmptyState: true)
            ]),
            const SizedBox(height: 50),
            Row(mainAxisAlignment: MainAxisAlignment.center, children: [
              ElevatedButton(
                  onPressed: _doPaymentFlow, child: const Text('Pay')),
            ]),
            ApplePayButton(
              paymentConfigurationAsset: 'payment_config.json',
              paymentItems: _paymentItems,
              style: ApplePayButtonStyle.black,
              type: ApplePayButtonType.buy,
              margin: const EdgeInsets.only(top: 15.0),
              onPaymentResult: onApplePayResult,
              loadingIndicator: const Center(
                child: CircularProgressIndicator(),
              ),
            ),
          ],
        )),
      ),
    );
  }
}
